[K32W0] SDK 2.6.13 additional changes

marius-alex-tache opened the pull request:
Changelog:

- Add support for Product Appearance.
- Various minor bug fixes.

Signed-off-by: Andrei Menzopol [andrei.menzopol@nxp.com](mailto:andrei.menzopol@nxp.com)
Signed-off-by: Damien Vagner [damien.vagner@nxp.com](mailto:damien.vagner@nxp.com)
Signed-off-by: Ethan Tan [Ethan.Tan@nxp.com](mailto:Ethan.Tan@nxp.com)
Signed-off-by: marius-alex-tache [marius.tache@nxp.com](mailto:marius.tache@nxp.com)

tanyue518 committed:
[K32W0] remove unused code in LowPowerHook.cpp file

marius-alex-tache committed:
[K32W] Make Validate public

Change access modifier of Validate to public, to be used by K32W0
for factory data validation after factory data OTA.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W0] Add CRC validation during factory data OTA

After factory data is copied into internal flash, call Validate to ensure
the CRC of factory data matches the CRC in the header. If the check fails,
then OTA is aborted and factory data is restored to previous version.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W0] Add additional check for provider pointer

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W] Expose SearchForId in factory data provider public API

Removed additional checks on tags with the assumption that factory data is
strictly checked at manufacturing time.
Applications can now use the public API to search the factory data section
for an id, e.g. in the case of custom factory data provider, where the app
has additional factory data information besides the default one.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W0] Refactor custom factory data provider

The example now leverages the actual factory data provider API to
search the factory data section for some custom ids.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W] Fix a corner case when switching the TLV processors

After the current block is processed, if the block size is 0, it means
that the processed data was a multiple of received BDX block size (e.g.
8 blocks of 1024 bytes were transferred). After state for selecting next
processor is reset, a request for fetching next data must be sent.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W] Add platform support for Product Appearance

Implement GetProductFinish and GetProductPrimaryColor defined in DeviceInstanceInfoProvider.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W] Remove maxLengths array

This array was used to store the maximum length of factory data fields.
It's no longer used, with the assumption that the factory data is strictly
checked in manufacturing.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W0] Remove usage of maxLength array

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

Commit:
[K32W0] use Encoding::HexToBytes to parse the ota encryption key

Commit:
[K32W0] Fix ICD parameters

Name of the ICD parameters were updated according with the latest stack updates.

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

Commit:
[K32W0] remove extra PWR_DisallowDeviceToSleep() which will cause unsleep after commissioning

marius-alex-tache committed:
[K32W] Revert removal of disallow to sleep and enclose it in specific tag

Only call disallow to sleep when the platform is K32W1.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W0] Update reference app readme files

This is a minor improvement to the building instructions.
It aims to clarify the usage of west SDK or package SDK.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

dvagner committed:
[K32W0] Change root node revision to 2

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=31382) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=31382) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: marius-alex-tache<br/>:white_check_mark: tanyue518<br/>:white_check_mark: andrei-menzopol<br/>:x: dvagner<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=31382) it.</sub>

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

restyled-commits committed:
Restyled by prettier-markdown

marius-alex-tache committed:
[K32W0] Fix gn check error

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

restyled-commits committed:
Restyled by gn

marius-alex-tache committed:
[K32W1] Fix gn check errors

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

restyled-commits committed:
Restyled by gn

marius-alex-tache committed:
[K32W1] Fix another gn error

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

restyled-commits committed:
Restyled by gn

marius-alex-tache commented on src/platform/nxp/k32w/k32w1/CHIPDevicePlatformConfig.h:
```diff
@@ -44,6 +44,8 @@
 #define CHIP_DEVICE_CONFIG_ENABLE_CHIP_TIME_SERVICE_TIME_SYNC 0
 // #define CHIP_DEVICE_CONFIG_PERSISTED_STORAGE_GLOBAL_EIDC_KEY 2
 
+#define CHIP_ADDRESS_RESOLVE_IMPL_INCLUDE_HEADER <lib/address_resolve/AddressResolve_DefaultImpl.h>
```
Ok, I will look into it.

marius-alex-tache replied:
The address resolve default config was now added to the platform device layer dependencies.

andy31415 reviewed (changes requested):
defining CHIP_ADDRESS_RESOLVE_IMPL_INCLUDE_HEADER as a config value seems suspect. It should be pulled using dependencies from address_resolve.

If platform depends on address resolve, that sounds like a layering problem - other platforms do not generally need this. I see stm32 and telink also doing odd stuff here, we have to fix those.

andy31415 commented:
> defining CHIP_ADDRESS_RESOLVE_IMPL_INCLUDE_HEADER as a config value seems suspect. It should be pulled using dependencies from address_resolve.
> 
> If platform depends on address resolve, that sounds like a layering problem - other platforms do not generally need this. I see stm32 and telink also doing odd stuff here, we have to fix those.

created #31392 and #31390 to fix the other platforms ... this define should really not reside anywhere near platform.

andy31415 commented on src/platform/nxp/k32w/k32w1/BUILD.gn:
```diff
@@ -106,4 +113,7 @@ static_library("k32w1") {
   public_deps += [ "${chip_root}/src/crypto" ]
+
+  public_configs =
+      [ "${chip_root}/src/lib/address_resolve:default_address_resolve_config" ]
```
this is not ok either - a platform cannot depend on address resolve, because address resolve will eventually depend on platform - platform code is one of the lowest layers we have.

I believe the problem may stem from changes I see in OTA code here - that code is in a `platform` folder however it includes code from `app` which seems a layering inversion. If you do need this (we had cases when platform needed to notify app of wifi changes for example) you have to break that dependency with delegates/callbacks or some other methods. Platform cannot call into app code directly.

marius-alex-tache replied:
I've removed the problematic code.

marius-alex-tache committed:
[K32W] Send a report before resetting the device during OTA

State-transition event from Downloading to Applying was not successfully sent
to a subscriber during OTA because the device would reset before actually sending
the ReportData message.

Added an explicit call to handle server shutting down, which will sync send all events.

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

restyled-commits committed:
Restyled by clang-format

andy31415 commented on src/platform/nxp/k32w/k32w1/BUILD.gn:
```diff
@@ -58,13 +58,20 @@ static_library("k32w1") {
   ]
 
+  public = [
+    "${chip_root}/src/app/reporting/reporting.h",
```
public on static libraries `Declare public header files for a target` according to `gn help public`.

however here src/app/reporting/reporting.h does not belong to k32w platform, so it seems very odd for the platform to declare "this is one of my public headers". What is going on here?

marius-alex-tache replied:
It's deprecated. I forgot to remove this.

marius-alex-tache committed:
[K32W0] Remove deprecated dependency

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

marius-alex-tache committed:
[K32W1] Remove deprecated dependency

Signed-off-by: marius-alex-tache <marius.tache@nxp.com>

restyled-commits committed:
Restyled by gn
